src/lender/test/reserve.t.sol

```diff
@@ -0,0 +1,134 @@
+        uint currencyAvailable = reserve.currencyAvailable();
+        
+        reserve.balance();
+
+        // assert currency was transferred from reserve to shelf
+        assertEq(currency.balanceOf(reserve_), safeSub(reserveBalance, borrowAmount));
+        assertEq(currency.balanceOf(shelf_), safeAdd(shelfBalance, borrowAmount));
+        assertEq(reserve.currencyAvailable(), safeSub(currencyAvailable, borrowAmount));
+    }
+
+    function testReserveBalanceRepay() public {
+        uint repayAmount = 100 ether;
+        bool requestWant = false;
+        // fund shelf with enough currency
+        currency.mint(shelf_, repayAmount);
+        // borrow action: shelf requests currency
+        shelf.setReturn("balanceRequest", requestWant, repayAmount);
+        // shelf approve reserve to take currency
+        shelf.doApprove(currency_, reserve_, repayAmount);
+        uint reserveBalance = currency.balanceOf(reserve_);
+        uint shelfBalance = currency.balanceOf(shelf_);
+        uint currencyAvailable = reserve.currencyAvailable();
+        reserve.balance();
+
+        // assert currency was transferred from shelf to reserve
+        assertEq(currency.balanceOf(reserve_), safeAdd(reserveBalance, repayAmount));
+        assertEq(currency.balanceOf(shelf_), safeSub(shelfBalance, repayAmount));
+        assertEq(reserve.currencyAvailable(), currencyAvailable);
+    }
+
+    function testFailBalancePoolInactive() public {
+        uint borrowAmount = 100 ether;
+        bool requestWant = true;
+        // fund reserve with enough currency
+        currency.mint(reserve_, 200 ether);
+        // borrow action: shelf requests currency
+        shelf.setReturn("balanceRequest", requestWant, borrowAmount);
+        // set max available currency
+        reserve.updateMaxCurrency(200 ether);
+        // deactivate pool
+        reserve.activatePool(false);
+        
+        reserve.balance();
+    }
+
+    function testFailBalanceBorrowAmountTooHigh() public {
+        uint borrowAmount = 100 ether;
+        bool requestWant = true;
+        // fund reserve with enough currency
+        currency.mint(reserve_, borrowAmount);
+        // borrow action: shelf requests too much currency
+        shelf.setReturn("balanceRequest", requestWant, safeMul(borrowAmount, 2));
+        // set max available currency
+        reserve.updateMaxCurrency(borrowAmount);
+        reserve.balance(); 
+    }
+
+    function testFailBalanceShelfNotEnoughFunds() public {
+        uint repayAmount = 100 ether;
+        bool requestWant = false;
+        // fund shelf with currency amount smaller then repay amount
+        currency.mint(shelf_, safeSub(repayAmount, 1));
+        // borrow action: shelf requests currency
+        shelf.setReturn("balanceRequest", requestWant, repayAmount);
+        // shelf approve reserve to take currency
+        shelf.doApprove(currency_, reserve_, repayAmount);
+        reserve.balance();
+    }
+
```

Comment:
Maybe we can add one more for reserve not enough funds and the edge case when it is equal